crates/chain/src/keychain/txout_index.rs

```diff
@@ -559,32 +559,75 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     }
 }
 
-fn range_descriptor_spks<'a, R>(
+fn range_descriptor_spks(
+    descriptor: Cow<'_, Descriptor<DescriptorPublicKey>>,
+    range: Range<u32>,
+) -> SpkIterator {
+    SpkIterator {
+        next_index: range.start,
+        end: range.end,
+        descriptor,
+        secp: Secp256k1::verification_only(),
+    }
+}
+
+#[derive(Clone)]
+pub struct SpkIterator<'a> {
+    next_index: u32,
+    end: u32,
     descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>,
-    range: R,
-) -> impl Iterator<Item = (u32, Script)> + Clone + Send + 'a
-where
-    R: Iterator<Item = u32> + Clone + Send + 'a,
-{
-    let secp = Secp256k1::verification_only();
-    let has_wildcard = descriptor.has_wildcard();
-    range
-        .into_iter()
-        // non-wildcard descriptors can only have one derivation index (0)
-        .take_while(move |&index| has_wildcard || index == 0)
-        // we can only iterate over non-hardened indices
-        .take_while(|&index| index <= BIP32_MAX_INDEX)
-        .map(
-            move |index| -> Result<_, miniscript::descriptor::ConversionError> {
-                Ok((
-                    index,
-                    descriptor
-                        .at_derivation_index(index)
-                        .derived_descriptor(&secp)?
-                        .script_pubkey(),
-                ))
-            },
-        )
-        .take_while(Result::is_ok)
-        .map(Result::unwrap)
+    secp: Secp256k1<bitcoin::secp256k1::VerifyOnly>,
+}
+
+impl<'a> SpkIterator<'a> {
+    // Creates a new script pubkey iterator starting at 0 from a descriptor
+    pub fn new(descriptor: Cow<'a, Descriptor<DescriptorPublicKey>>) -> Self {
```

Comment:
It'll be good to have another constructor where you can specify start index and an optional end index.

Reply:
Completed.